game-core/src/main/java/games/strategy/triplea/delegate/BattleTracker.java

```diff
@@ -156,9 +156,8 @@ boolean didAllThesePlayersJustGoToWarThisTurn(
   private boolean didThesePlayersJustGoToWarThisTurn(final PlayerId p1, final PlayerId p2) {
     // check all relationship changes that are p1 and p2, to make sure that oldRelation is not war,
     // and newRelation is war
-    for (final Tuple<Tuple<PlayerId, PlayerId>, Tuple<RelationshipType, RelationshipType>> t :
-        relationshipChangesThisTurn) {
-      final Tuple<PlayerId, PlayerId> players = t.getFirst();
+    for (final var tuple : relationshipChangesThisTurn) {
```

Comment:
`tuple` probably could use a rename; This is a scary bit of code, it might even be worth leaving alone and waiting for a more significant refactor to fix it up.
Failing that, a `var` in foreach loop I think raises question if the lambda foreach variant shoudl be used, eg:
`relationshipChangesThisTurn.forEach(tuple -> ...)`.

Overall it's a code smell for procedural code. The refactoring fix I think is to make `relationshipChangesThisTurn` a first class object. In which case the for loop is extracted entirely to the new object and we get something like this: `return relationshipChangesThis.arePlayersAtWar(p1, p2)`. In that case the entire private method is probably extracted to the new object since the method is essentially just this for loop.

Reply:
That's the piece of code I was refferring to in another PR.
And I agree with you, but for now the only goal is to make the code more readable, and I think you can agree that the existing code isn't really better.

Reply:
While the existing code was ugly, the type information is there. `var` is good IMO when the type information is redundant and obvious, eg:
- `List playerList = new ArrayList<>();`
- `RelationshipTracker relationshipTracker = new RelationshipTracker()`

Type information is helpful, so while harder to read with the nasty Tuple mess, it does add information of what data is there. Losing that requires a reader to do a lookup of the definition of the `relationshipChangesThisTurn` variable and cross-reference its type with the usage, which I think actually detracts from understandability (can't just understand this code by looking at it, a cross-reference to other locations is an undersatndability smell)

Reply:
I do notice now more clearly you added lines like: `  final Tuple<PlayerId, PlayerId> players = tuple.getFirst();`

Perhaps that is enough to help and mitigates the typing concern.